crmsh/ui_resource.py

```diff
@@ -497,11 +498,54 @@ def do_constraints(self, context, resource):
 
     @command.wait
     @command.completers(compl.resources, _attrcmds, compl.nodes)
-    def do_failcount(self, context, rsc, cmd, node, value=None):
+    def do_failcount(self, context, rsc, cmd, node, value=None, operation=None, interval=None):
         """usage:
-        failcount <rsc> set <node> <value>
+        failcount <rsc> set <node> <value> [operation] [interval]
         failcount <rsc> delete <node>
         failcount <rsc> show <node>"""
+        if cmd == "set":
+            nodeid = utils.get_nodeid_from_name(node)
+            if nodeid is None:
+                context.fatal_error("Node {} not in this cluster".format(node))
+
+            # query the current failcount status
+            query_cmd = "cibadmin -Ql --xpath '/cib/status/node_state[@id='{}']'".format(nodeid)
+            rc, out, err = utils.get_stdout_stderr(query_cmd)
+            if rc != 0:
+                context.fatal_error(err)
+
+            # try to get failcount dict {operation:interval}
+            import re
+            failcount_res = re.findall(r'fail-count-{}#(.*)_([0-9]+)'.format(rsc), out)
+            if not failcount_res:
+                context.fatal_error("No failcount on node {}".format(node))
+            failcount_dict = dict(failcount_res)
+
+            # validate for operation and interval
+            if operation and operation not in failcount_dict.keys():
+                context.fatal_error("Should specify operation between \"{}\"".format(' '.join(failcount_dict.keys())))
+            if (operation and interval) and (operation, interval) not in failcount_res:
+                context.fatal_error("Should specify (operation, interval) between {}".format(failcount_res))
+
+            # just one failcount entry
+            if len(failcount_res) == 1:
+                operation = failcount_res[0][0]
+                interval = failcount_dict[operation]
+                rsc = '{}#{}_{}'.format(rsc, operation, interval)
+
+            # multiple failcount entries for this resource and node
+            if len(failcount_res) > 1:
+                if operation and interval:
+                    rsc = '{}#{}_{}'.format(rsc, operation, interval)
+                else:
+                    # using '-P' option of 'crm_attribute' command
+                    cmd = "set_p"
+                    if operation:
+                        op_interval_str = '|'.join(['{}_{}'.format(operation, inter) for op, inter in failcount_res if op==operation])
+                    else:
+                        op_interval_str = '|'.join(['{}_{}'.format(op, inter) for op, inter in failcount_res])
+                    rsc = '{}#({})'.format(rsc, op_interval_str)
+
```

Comment:
I'd go with some logic like the following. Although it looks a little sophisticated, but it'd be very friendly ;-)
```
if operation == None and interval == None:
    if len(existing_failcount_entries) == 0:
        if rsc_has_only_one_monitor_configured(rsc):
            # Set fail-count for monitor with the _configured_ interval (don't forget *1000)
        else:
            # Error: Please specify operation and interval
    elif len(existing_failcount_entries) == 1:
        # Set the only existing fail-count entry to value
    elif len(existing_failcount_entries) > 1:
        if value == 0:
            # Set all the existing fail-count entries to 0
        else:
            # Error: Please specify operation and interval
elif operation != None:
    if not is_valid_op(rsc, operation): #start/stop/monitor, promote/demote if rsc is promotable
        # Error: Please specify a valid operation
    elif interval == None:
        if len(failcount_entries_matching_operation) == 0:
            if is_valid_one_shot_op(rsc, operation):  #start/stop, promote/demote if rsc is promotable 
                # Set fail-count for the operation with interval 0
            elif operation == "monitor" and rsc_has_only_one_monitor_configured(rsc):
                # Set fail-count for monitor with the _configured_ interval
            else:
                # Error: Please specify operation and interval
        elif len(failcount_entries_matching_operation) == 1:
            # Set the only fail-count entry matching the operation to value
        elif len(failcount_entries_matching_operation) > 1:
            if value == 0:
                # Set all the fail-counts matching the operation to 0
            else:
                #  Error: Please specify operation and interval
    else: # interval != None
        if len(failcount_entries_matching_operation_interval) > 0:
            # Set fail-count for operation with the specified interval anyway
        elif is_valid_one_shot_op(rsc, operation) and interval != 0:
                # Error: interval should be 0 or omitted for operation %s
        elif operation == "monitor" and not rsc_has_monitor_configured_with_interval(rsc, interval):
                # Error: Please specify a configured monitor interval for rsc
        else:
                # Set fail-count with the specified operation and interval
else: # operation == None and interval != None -- is it possible?
    ...
```    
Another thing is, we accept interval in "seconds" specified at command line interface, but we need to convert it to "ms" as suffix of the fail-count entry name.

Reply:
@gao-yan Thanks for your suggestions. 
I might keep my opinion.
1) This bugs' requirement from our customer is simple: 
`Keep this ticket to track "crm resource failcount <rsc> set <node> 0" fix.`
That will work if cib has one or more failcount entries;
In doc, I give the syntax like:
`failcount <rsc> set <node> <value> [operation] [interval]`
That will also work if user input the `<operation> and <interval>`, I will always do validation work if they were not null, to lead user give the right and exists `<operation> and <interval>` pair.
2) I think it make no sense to create a new fail-count entry if this fail-count for this operation not exist, and that is not the requirement from customer. So if there is no fail-count entry exists while running this command, I will give error like:
`No failcount on node {} for resource {}`